controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -728,27 +768,143 @@ private long findSegmentSplitsMerges(List<Segment> referenceSegmentsList, List<S
+                            final Data<T> updated = new Data<>(updatedTable, historyTable.getVersion());
+
+                            return addSealedSegmentsToRecord(sealedSegments)
+                                    .thenCompose(x -> updateHistoryTable(updated))
+                                    .whenComplete((r, e) -> {
+                                        if (e != null) {
+                                            log.warn("{}/{} attempt to complete epoch transition for epoch {}. {}", scope, name, activeEpoch,
+                                                    e.toString());
+                                        } else {
+                                            log.debug("{}/{} epoch transition complete, index and history tables updated for epoch {}.",
+                                                    scope, name, activeEpoch);
+                                        }
+                                    });
+                        }));
+    }
+
+    @Override
+    public CompletableFuture<Void> rollingTxnNewSegmentsCreated(Map<Long, Long> sealedTxnEpochSegments, int transactionEpoch, long time) {
+        return checkState(state -> state.equals(State.COMMITTING_TXN) || state.equals(State.SEALING))
+        .thenCompose(v -> addSealedSegmentsToRecord(sealedTxnEpochSegments)
+                .thenCompose(x -> getActiveEpoch(true))
+                .thenCompose(activeEpoch -> rollingTxnAddNewDuplicateEpochs(transactionEpoch, time)));
+    }
+
+    private CompletableFuture<Void> rollingTxnAddNewDuplicateEpochs(final int transactionEpoch, final long time) {
+        return getHistoryIndexFromStore()
+                .thenCompose(historyIndex -> getHistoryTableFromStore()
+                        .thenCompose(historyTable -> {
+                            final int activeEpoch = TableHelper.getActiveEpoch(historyIndex.getData(), historyTable.getData()).getEpoch();
+                            final HistoryRecord lastRecord = HistoryRecord.readLatestRecord(historyIndex.getData(), historyTable.getData(),
+                                    false).get();
+                            int newEpoch = activeEpoch + 2;
+
+                            // idempotent check
+                            if (lastRecord.getEpoch() > activeEpoch) {
+                                boolean idempotent = lastRecord.isPartial() && lastRecord.getEpoch() == newEpoch &&
+                                        lastRecord.getReferenceEpoch() == activeEpoch;
+
+                                if (idempotent) {
+                                    HistoryRecord previous = HistoryRecord.fetchPrevious(lastRecord, historyIndex.getData(),
+                                            historyTable.getData()).get();
+                                    idempotent &= previous.getReferenceEpoch() == transactionEpoch;
+                                }
+
+                                if (idempotent) {
```

Comment:
These two consecutive if blocks are a bit odd. If the first isn't executed, then the second is not executed either. If the first executed, then maybe the second is also executed. It feels like this second i block should be incorporated into the first. I can't make much sense of the else block, it feels like a safety check and could remain where it is without the else even.

In any case, please revisit this logic.